evalml/model_understanding/_partial_dependence.py

```diff
@@ -0,0 +1,251 @@
+
+
+def _grid_from_X(X, percentiles, grid_resolution, custom_range):
+    """Create cartesian product of all the columns of input dataframe X.
+
+    Args:
+        X (pd.DataFrame): Input data
+        percentiles (tuple float): Percentiles to use as endpoints of the grid
+            for each feature.
+        grid_resolution (int): How many points to interpolate between percentiles.
+        custom_range (dict[str, np.ndarray]): Mapping from column name in X to
+            range of values to use in partial dependence. If custom_range is specified,
+            the percentile + interpolation procedure is skipped and the values in custom_range
+            are used.
+
+    Returns:
+        pd.DataFrame: Cartesian product of input columns of X.
+    """
+    values = []
+    for feature in X.columns:
+        if feature in custom_range:
+            # Use values in the custom range
+            feature_range = custom_range[feature]
+            if not isinstance(feature_range, (np.ndarray, pd.Series)):
+                feature_range = np.array(feature_range)
+            if feature_range.ndim != 1:
+                raise ValueError(
+                    "Grid for feature {} is not a one-dimensional array. Got {}"
+                    " dimensions".format(feature, feature_range.ndim)
+                )
+            axis = feature_range
+        else:
+            uniques = np.unique(X.loc[:, feature])
+            if uniques.shape[0] < grid_resolution:
+                # feature has low resolution use unique vals
+                axis = uniques
+            else:
+                # create axis based on percentiles and grid resolution
+                emp_percentiles = mquantiles(
+                    X.loc[:, feature], prob=percentiles, axis=0
+                )
+                if np.allclose(emp_percentiles[0], emp_percentiles[1]):
+                    raise ValueError(
+                        "percentiles are too close to each other, "
+                        "unable to build the grid. Please choose percentiles "
+                        "that are further apart."
+                    )
+                axis = np.linspace(
+                    emp_percentiles[0],
+                    emp_percentiles[1],
+                    num=grid_resolution,
+                    endpoint=True,
+                )
+        values.append(axis)
+
+    return _cartesian(values), values
+
+
+def _cartesian(arrays):
+    """Create cartesian product of elements of arrays list.
+
+    Stored in a dataframe to allow mixed types like dates/str/numeric.
+
+    Args:
+        arrays (list(np.ndarray)): Arrays.
+
+    Returns:
+        pd.DataFrame: Cartesian product of arrays.
+    """
+    arrays = [np.asarray(x) for x in arrays]
+    shape = (len(x) for x in arrays)
+
+    ix = np.indices(shape)
+    ix = ix.reshape(len(arrays), -1).T
+
+    out = pd.DataFrame()
+
+    for n, arr in enumerate(arrays):
+        out[n] = arrays[n][ix[:, n]]
+
+    return out
```

Comment:
This seems to be the same as https://github.com/scikit-learn/scikit-learn/blob/844b4be24d20fc42cc13b957374c718956a0db39/sklearn/utils/extmath.py#L655 except we return a dataframe, and since it's a public sklearn method, we could just import it--whatcha think? Also totally down to take their impl 😂

Reply:
This is a great idea.  I know as we first tredged through partial dependence that we borrowed a lot.  Perhaps a bit more from some private methods than I would like, but it was necessary.  If we can refactor to use their public methods, that's great.

Reply:
@angela97lin Great point. Originally I wanted to use their method but the problem is that numpy arrays cannot handle mixed-types very well. So if we want to have a grid of categoricals and datetimes, the conversion storing it in a numpy array won't really work.

![image](https://user-images.githubusercontent.com/41651716/135173220-b45fbc4b-164e-4007-b1e5-3b67cc64ad1c.png)

There may be a way around it I'm not seeing (maybe [this](https://numpy.org/doc/stable/user/basics.rec.html)) but IMO that's a nice to have?